File: js/src/core/filterE2E.ts
import { COCO_CLASSES } from "./classes";
import type { ArrayDetections, Detection, FilterE2EOptions } from "./types";

function at(arr: number[], i: number): number {
  // biome-ignore lint/style/noNonNullAssertion: index is always in-bounds at call sites
  return arr[i]!;
}

function atF32(arr: Float32Array, i: number): number {
  // biome-ignore lint/style/noNonNullAssertion: index is always in-bounds at call sites
  return arr[i]!;
}

export function filterE2E(
  output: Float32Array,
  shape: readonly number[],
  opts: FilterE2EOptions = {},
): Detection[] | ArrayDetections {
  const { conf = 0.25, classes, minArea, format = "dict", strict = false } = opts;
  if (conf < 0 || conf > 1) throw new Error(`conf must be in [0, 1]; got ${conf}`);

  let K: number;
  let lastDim: number;
  if (shape.length === 3) {
    if (shape[0] !== 1) throw new Error(`batched decode not supported; got batch=${shape[0]}`);
    K = shape[1] as number;
    lastDim = shape[2] as number;
  } else if (shape.length === 2) {
    K = shape[0] as number;
    lastDim = shape[1] as number;
  } else {
    throw new Error(`expected (K,6) or (1,K,6); got rank ${shape.length}`);
  }
  if (lastDim !== 6) throw new Error(`filter_e2e expects last dim = 6; got ${lastDim}`);

  const boxes: number[] = [];
  const scores: number[] = [];
  const cls: number[] = [];
  const idxs: number[] = [];

  const allowlist = classes ? new Set(classes) : null;
  if (allowlist) {
    for (const c of allowlist) {
      if (c < 0 || c >= COCO_CLASSES.length) {
        throw new Error(`classes allowlist out of range: ${c}`);
      }
    }
  }

  for (let i = 0; i < K; i++) {
    const off = i * 6;
    const x1 = atF32(output, off);
    const y1 = atF32(output, off + 1);
    const x2 = atF32(output, off + 2);
    const y2 = atF32(output, off + 3);
    const confI = atF32(output, off + 4);
    const cid = atF32(output, off + 5) | 0;

    const finite =
      Number.isFinite(x1) &&
      Number.isFinite(y1) &&
      Number.isFinite(x2) &&
      Number.isFinite(y2) &&
      Number.isFinite(confI);
    if (!finite) {
      if (strict) throw new Error("output contains NaN/Inf");
      continue;
    }
    if (confI < conf) continue;
    if (allowlist && !allowlist.has(cid)) continue;
    if (cid < 0 || cid >= COCO_CLASSES.length) {
      if (strict) throw new Error(`class_id out of range: ${cid}`);
      continue;
    }
    if (minArea !== undefined) {
      const area = (x2 - x1) * (y2 - y1);
      if (area < minArea) continue;
    }
    boxes.push(x1, y1, x2, y2);
    scores.push(confI);
    cls.push(cid);
    idxs.push(i);
  }

  const order = scores
    .map((_, i) => i)
    .sort((a, b) => {
      const sd = at(scores, b) - at(scores, a);
      if (sd !== 0) return sd;
      const cd = at(cls, a) - at(cls, b);
      if (cd !== 0) return cd;
      return at(idxs, a) - at(idxs, b);
    });

  const N = order.length;
  const outBoxes = new Float32Array(N * 4);
  const outScores = new Float32Array(N);
  const outClasses = new Int32Array(N);
  for (let k = 0; k < N; k++) {
    const j = at(order, k);
    outBoxes[k * 4] = at(boxes, j * 4);
    outBoxes[k * 4 + 1] = at(boxes, j * 4 + 1);
    outBoxes[k * 4 + 2] = at(boxes, j * 4 + 2);
    outBoxes[k * 4 + 3] = at(boxes, j * 4 + 3);
    outScores[k] = at(scores, j);
    outClasses[k] = at(cls, j);
  }

  if (format === "arrays") {
    return { boxes: outBoxes, scores: outScores, classes: outClasses };
  }
  if (format === "dict") {
    const dets: Detection[] = [];
    for (let k = 0; k < N; k++) {
      const c = atF32(outClasses as unknown as Float32Array, k) | 0;
      dets.push({
        box: [
          atF32(outBoxes, k * 4),
          atF32(outBoxes, k * 4 + 1),
          atF32(outBoxes, k * 4 + 2),
          atF32(outBoxes, k * 4 + 3),
        ],
        score: atF32(outScores, k),
        class: c,
        label: COCO_CLASSES[c] as string,
      });
    }
    return dets;
  }
  throw new Error(`unknown format: ${format}`);
}

File: python/src/yolo26_kit/core/filter_e2e.py
"""Filter helper for YOLO26 end-to-end (default) ONNX export output."""
from __future__ import annotations

from collections.abc import Iterable
from typing import Literal

import numpy as np

from .types import COCO_CLASSES, Detection

_FormatT = Literal["dict", "arrays"]


def filter_e2e(
    output: np.ndarray,
    conf: float = 0.25,
    classes: Iterable[int] | None = None,
    min_area: float | None = None,
    format: _FormatT = "dict",
    *,
    strict: bool = False,
) -> list[Detection] | dict[str, np.ndarray]:
    """Filter and format the (N, K, 6) e2e YOLO26 output.

    See spec/decode.md Algorithm A.
    """
    if not 0.0 <= conf <= 1.0:
        raise ValueError(f"conf must be in [0, 1]; got {conf}")

    arr = np.asarray(output)
    if arr.ndim == 2:
        pass
    elif arr.ndim == 3:
        if arr.shape[0] != 1:
            raise ValueError(
                f"batched decode not supported in v1 — got batch={arr.shape[0]}; "
                "call per-batch-item",
            )
        arr = arr[0]
    else:
        raise ValueError(f"expected (K, 6) or (1, K, 6); got shape {arr.shape}")

    if arr.shape[-1] != 6:
        raise ValueError(
            f"filter_e2e expects last dim = 6 [x1,y1,x2,y2,conf,cls]; got {arr.shape[-1]}",
        )

    arr = arr.astype(np.float32, copy=False)
    boxes = arr[:, 0:4]
    scores = arr[:, 4]
    classes_arr = arr[:, 5].astype(np.int32)

    if not np.isfinite(arr).all():
        if strict:
            raise ValueError("output contains NaN/Inf")
        finite_mask = np.isfinite(arr).all(axis=1)
    else:
        finite_mask = np.ones(arr.shape[0], dtype=bool)

    mask = finite_mask & (scores >= conf)

    # Class-id range validation: drop or error on out-of-range class ids,
    # so downstream COCO_CLASSES lookup never goes out-of-bounds.
    if classes_arr.size:
        valid_cls = (classes_arr >= 0) & (classes_arr < len(COCO_CLASSES))
        if not valid_cls.all():
            if strict:
                raise ValueError("class_id out of range")
            mask &= valid_cls

    if classes is not None:
        allowlist = np.fromiter(classes, dtype=np.int32)
        if allowlist.size:
            if (allowlist < 0).any() or (allowlist >= len(COCO_CLASSES)).any():
                raise ValueError(f"classes allowlist out of range: {allowlist.tolist()}")
            mask &= np.isin(classes_arr, allowlist)

    if min_area is not None:
        widths = boxes[:, 2] - boxes[:, 0]
        heights = boxes[:, 3] - boxes[:, 1]
        mask &= (widths * heights) >= min_area

    boxes = boxes[mask]
    scores = scores[mask]
    classes_arr = classes_arr[mask]

    # Sort: descending score, stable tiebreak by (class asc, source idx asc).
    # np.lexsort uses ASCENDING; trailing key is primary. Negate score for desc.
    src_idx = np.arange(scores.shape[0], dtype=np.int64)
    order = np.lexsort((src_idx, classes_arr, -scores))
    boxes = boxes[order]
    scores = scores[order]
    classes_arr = classes_arr[order]

    if format == "arrays":
        return {
            "boxes": np.ascontiguousarray(boxes, dtype=np.float32),
            "scores": np.ascontiguousarray(scores, dtype=np.float32),
            "classes": np.ascontiguousarray(classes_arr, dtype=np.int32),
        }
    if format == "dict":
        out: list[Detection] = [
            {
                "box": [float(b[0]), float(b[1]), float(b[2]), float(b[3])],
                "score": float(s),
                "class": int(c),
                "label": COCO_CLASSES[int(c)],
            }
            for b, s, c in zip(boxes, scores, classes_arr, strict=True)
        ]
        return out
    raise ValueError(f"unknown format: {format!r}; expected 'dict' or 'arrays'")
